Fixed the working of order history pagination (#4599-Order-History-Pagination-Not-Working)

AishwaryShrivastav opened the pull request:
…

Called the 'getOrdersHistory' function on 'onBottomScroll' and changed the mutation to also save the recent orders.

### Related Issues
<!--  Put related issue number which this PR is closing. For example #123 -->

closes #4599 

### Short Description and Why It's Useful
<!-- describe in a few words what is this Pull Request changing and why it's useful -->
the getOrdersHistory, which previously used to fetch all the orders, after introducing pagination support in "getOrdersHistory", it only fetches 20 orders, which is its default value.

Still, the order history page calls it only once, trying to fetch all the results, and on those results, it applies lazyLoadOrdersOnScroll.

Due to which the user can not see more than 20 orders.


### Screenshots of Visual Changes before/after (if There Are Any)
<!-- if you made any changes in the UI layer please provide before/after screenshots -->

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance
<!-- Please get familiar with following info -->

- [X] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/DivanteLtd/vue-storefront?pullRequest=4632) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/DivanteLtd/vue-storefront?pullRequest=4632) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/DivanteLtd/vue-storefront?pullRequest=4632) it.</sub>

AishwaryShrivastav commented:
Many thanks to @dixitdeepak, @swapnilmmane, @mohammad-k8 and @gibkigonzo for the help and discussion.

dixitdeepak commented on core/modules/user/store/mutations.ts:
```diff
@@ -24,7 +24,13 @@ const mutations: MutationTree<UserState> = {
   [types.USER_ORDERS_HISTORY_LOADED] (state, ordersHistory) {
-    state.orders_history = ordersHistory
+    if (state.orders_history) {
+      ordersHistory.items.forEach(item => {
```
I think instead of forEach you can use Object.assign

AishwaryShrivastav commented:
Many thanks @dixitdeepak for suggesting all the improvements, 
according to them, moved the logic to actions instead of mutations, added condition to clear the orders on logout and improved the pagination check.

AishwaryShrivastav commented:
Here is a video on this PR for better understanding of the changes https://www.youtube.com/watch?v=rjjcb3-AfII&feature=youtu.be

pschaub commented:
Please be sure that if you trigger `refreshOrdersHistory` multiple times that the order is not duplicated in the list. Each order should be only once in the order history even if you trigger `refreshOrdersHistory` multiple times.

Fifciu commented on core/modules/order/components/UserOrdersHistory.ts:
```diff
@@ -1,13 +1,14 @@
     return {
       pagination: {
-        perPage: 10,
+        perPage: 20,
```
Why changed to 20?

AishwaryShrivastav replied:
Changed it back to 10.

Fifciu commented on core/modules/order/components/UserOrdersHistory.ts:
```diff
@@ -26,6 +27,13 @@ export default {
     onBottomScroll () {
       ++this.pagination.current
-    }
+      let total_count = this.$store.state.user.orders_history.total_count ? this.$store.state.user.orders_history.total_count : 0
```
In PWA, we prefer camelCase - please use it. In the object we have snake_case only because it comes directly from M2. Also please change it to const as it never changes

AishwaryShrivastav replied:
modified as per your suggestion

Fifciu commented on core/modules/order/components/UserOrdersHistory.ts:
```diff
@@ -26,6 +27,13 @@ export default {
       ++this.pagination.current
-    }
+      let total_count = this.$store.state.user.orders_history.total_count ? this.$store.state.user.orders_history.total_count : 0
+      if (Math.ceil(total_count / this.pagination.perPage) >= this.pagination.current) {
```
What about changing it to:
```js
const isLastPage = this.pagination.current > Math.ceil(total_count / this.pagination.perPage);
if (!isLastPage) {
```

Then everyone would understand what this condition means even with 0 information about Math.ceil and these variables

AishwaryShrivastav replied:
done

Fifciu commented on core/modules/user/store/actions.ts:
```diff
@@ -257,12 +258,20 @@ const actions: ActionTree<UserState, RootState> = {
+      EventBus.$emit('user-after-loaded-orders', orders)
     }
 
     if (!resolvedFromCache) {
```
Please add return before this `Promise.resolve`

AishwaryShrivastav replied:
modified it to return Promise.resolve....

Fifciu commented on core/modules/user/store/actions.ts:
```diff
@@ -257,12 +258,20 @@ const actions: ActionTree<UserState, RootState> = {
+      if (oldOrders && orders.items) {
+        orders.items = [
+          ...oldOrders,
+          ...resp.result.items
```
Why not `...order.items`?

AishwaryShrivastav replied:
nice catch, modified

Fifciu commented on core/modules/user/store/actions.ts:
```diff
@@ -257,12 +258,20 @@ const actions: ActionTree<UserState, RootState> = {
     if (resp.code === 200) {
-      commit(types.USER_ORDERS_HISTORY_LOADED, resp.result) // this also stores the current user to localForage
-      EventBus.$emit('user-after-loaded-orders', resp.result)
+      let oldOrders = getters.getOrdersHistory;
```
Let's make it const. I am pretty sure both of them could be `const`

AishwaryShrivastav replied:
Modified both to const

Fifciu reviewed (changes requested):
Please apply post-CR fixes, retest manually, and update the changelog

Fifciu commented on core/modules/user/store/actions.ts:
```diff
@@ -257,16 +258,24 @@ const actions: ActionTree<UserState, RootState> = {
     }
   },
-  async refreshOrdersHistory ({ commit }, { resolvedFromCache, pageSize = 20, currentPage = 1 }) {
+  async refreshOrdersHistory ({ commit, getters }, { resolvedFromCache, pageSize = 20, currentPage = 1 }) {
```
I can see now that here we have pageSize=20, and inside mixin we have 10. Is `10` properly applied in each place?

Fifciu reviewed (changes requested):
I see this PR is based and points to bad branch. It should be `hotfix/v1.12.3` not `master`

Fifciu commented:
https://github.com/vuestorefront/vue-storefront/pull/5527
